src/GraphQL/Types/ArgumentInformation.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading;
+
+namespace GraphQL.Types
+{
+    public abstract class ArgumentInformation
+    {
+        protected ArgumentInformation(ParameterInfo parameterInfo, Type sourceType, FieldType fieldType, TypeInformation typeInformation, LambdaExpression? expression)
+        {
+            ParameterInfo = parameterInfo ?? throw new ArgumentNullException(nameof(parameterInfo));
+            FieldType = fieldType ?? throw new ArgumentNullException(nameof(fieldType));
+            SourceType = sourceType ?? throw new ArgumentNullException();
+            TypeInformation = typeInformation ?? throw new ArgumentNullException(nameof(typeInformation));
+            Expression = expression;
+        }
+
+        protected ArgumentInformation(ParameterInfo parameterInfo, Type sourceType, FieldType fieldType, TypeInformation typeInformation)
+            : this(parameterInfo, sourceType, fieldType, typeInformation, null)
+        {
+            if (parameterInfo.ParameterType == typeof(IResolveFieldContext))
+            {
+                Expression<Func<IResolveFieldContext, IResolveFieldContext>> expr = x => x;
+                Expression = expr;
+            }
+            else if (parameterInfo.ParameterType == typeof(CancellationToken))
+            {
+                Expression<Func<IResolveFieldContext, CancellationToken>> expr = x => x.CancellationToken;
+                Expression = expr;
+            }
+        }
+
+        public ParameterInfo ParameterInfo { get; }
+
+        public Type SourceType { get; }
+
+        public FieldType FieldType { get; }
```

Comment:
`SourceType` is useful if the attribute logic needs to interact with `context.Source`.

`FieldType` may not really be needed.  It provides attributes with information about the field to which the argument is being added.  I have to say I can't think of a scenario where it would be useful, so we could remove it.